hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/RefreshMetaProcedure.java

```diff
@@ -0,0 +1,423 @@
+        LOG.info("Meta table update completed successfully");
+      } else {
+        LOG.info("No update needed - meta table is in sync with backing storage");
+      }
+
+      setNextState(RefreshMetaState.REFRESH_META_COMPLETE);
+      return Flow.HAS_MORE_STATE;
+    } catch (IOException ioe) {
+      LOG.error("Failed to update the hbase:meta table", ioe);
+      throw ioe;
+    }
+  }
+
+  private Flow executeComplete() {
+    LOG.info("RefreshMetaProcedure completed successfully");
+    return Flow.NO_MORE_STATE;
+  }
+
+  /**
+   * Compares the current regions with the latest regions and updates the meta table if necessary.
+   */
+  void compareAndUpdateRegions(List<RegionInfo> current, List<RegionInfo> latest,
+    Connection connection) throws IOException {
+    List<Put> puts = new ArrayList<>();
+    List<Delete> deletes = new ArrayList<>();
+    Set<RegionInfo> currentSet = new HashSet<>(current);
+    Set<RegionInfo> latestSet = new HashSet<>(latest);
+
+    // Find regions to add (present in latest but not in current)
+    for (RegionInfo ri : latest) {
+      if (!currentSet.contains(ri)) {
+        puts.add(MetaTableAccessor.makePutFromRegionInfo(ri));
+        LOG.debug("Adding the region to meta: {}", ri.getRegionNameAsString());
+      }
+    }
+
+    // Find regions to remove (present in current but not in latest)
+    for (RegionInfo ri : current) {
+      if (!latestSet.contains(ri)) {
+        deletes.add(MetaTableAccessor.makeDeleteFromRegionInfo(ri,
+          EnvironmentEdgeManager.currentTime()));
+        LOG.debug("Removing region from meta: {}", ri.getRegionNameAsString());
+      }
+    }
+
+    // Find regions to update (same region name but different boundaries)
+    for (RegionInfo latestRegion : latest) {
+      if (currentSet.contains(latestRegion)) {
+        RegionInfo currentRegion = current.stream()
+          .filter(c -> c.getRegionNameAsString().equals(latestRegion.getRegionNameAsString()))
+          .findFirst()
+          .orElse(null);
+
+        if (currentRegion != null && hasBoundaryChanged(currentRegion, latestRegion)) {
+          puts.add(MetaTableAccessor.makePutFromRegionInfo(latestRegion));
+          LOG.debug("Adding a put to update region boundaries in meta: {}",
+            latestRegion.getRegionNameAsString());
+        }
+      }
+    }
+
+    if (!puts.isEmpty()) {
+      LOG.info("Adding/updating {} regions in meta table", puts.size());
+      executeBatchedPuts(connection, puts);
+    }
+
+    if (!deletes.isEmpty()) {
+      LOG.info("Removing {} regions from meta table", deletes.size());
+      executeBatchedDeletes(connection, deletes);
+    }
+
+    if (puts.isEmpty() && deletes.isEmpty()) {
+      LOG.info("No changes needed in meta table");
+    }
+  }
+
+  private void executeBatchedPuts(Connection connection, List<Put> puts) throws IOException {
+    for (int i = 0; i < puts.size(); i += CHUNK_SIZE) {
+      int end = Math.min(puts.size(), i + CHUNK_SIZE);
```

Comment:
You can partition a List easily with Guava or Apache Commons libraries.

(Guava)
```java
List<List<Put>> subSets = Lists.partition(puts, CHUNK_SIZE);
```

https://www.baeldung.com/java-list-split